src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceDescriptor.cs

```diff
@@ -72,6 +72,24 @@ private ServiceDescriptor(Type serviceType, ServiceLifetime lifetime)
         /// <inheritdoc />
         public Func<IServiceProvider, object> ImplementationFactory { get; }
 
+        public Type GetImplementationType()
+        {
+            if (ImplementationType != null)
+            {
+                return ImplementationType;
+            }
+            else if (ImplementationInstance != null)
+            {
+                return ImplementationInstance.GetType();
+            }
+            else if (ImplementationFactory != null)
+            {
+                return ImplementationFactory.GetType().GenericTypeArguments[1];
```

Comment:
Could you throw in a `Debug.Assert(ImplementationFactory.GetType().GenericTypeArguments.Length == 2)`?